EspressoSystems/nitro-espresso-integration
# Reimplement escape hatch

Closes #221 

### This PR:
Adds an escape hatch into the sovereign sequencer. When the light client reports that hotshot is down, this escape hatch will switch back to central using the centralized sequencer for the rollup.

### This PR does not:
Modify the portion of the STF present in `cmd/replay/main.go`.  It's a bit unclear to me what we need to be doing here now. With this new implementation, sometimes the data we send to espresso propagates slowly enough to cause `wavmio.isHotshotLive()` to fail which is causing validation to break.

### Key places to review:
The code in `hotshot_monitor.go`, changes in `batch_poster.go`, changes in `transaction_streamer.go` as well as changes to `execution/gethexec/sequencer.go`

#### Behavior immediately after the escape hatch closes.
During the E2E test it can be seen that the `batch_poster` and `transaction_streamer` have requests that fail several times right after the escape hatch closes. I believe this is a result of the requests happening before data (very recently ) submitted to the espresso network becomes available. I don't think this is a correctness issue as these components seem to unblock themselves, but this delay wasn't a previous behavior. It is at the very least a pollution of the logs.

### How to test this PR:
Running the E2E test with the `E2E_SKIP_ESCAPE_HATCH_TEST` env var not popluated.

The test will show the `HotShotMonitor` listen for espresso liveness during the startup of the espresso network. When the espresso network has advanced to a position that the light client will declare as live, the escape hatch will close and sequence transactions in espresso mode. After the normal E2E test, if the `E2E_SKIP_ESCAPE_HATCH_TEST` environment variable is empty, the light client mock will be used to pause the hotshot height. This causes the escape hatch to open so we can sequence transactions in espresso mode. Finally, we unfreeze the hotshot height with the mock light client and ensure that we validate transactions successfully.
